server_test.go

```diff
@@ -259,6 +259,99 @@ func TestServingTLS(t *testing.T) {
 	}
 }
 
+// TestServingTLSConnectionState tests that we only can access
+// tls.ConnectionState under a DNS query handled by a TLS DNS server.
+// This test will sequentially create a TLS, UDP and TCP server, attach a custom
+// handler which will set a testing error if tls.ConnectionState is available
+// when it is not expected, or the other way around.
+func TestServingTLSConnectionState(t *testing.T) {
+
+	handlerResponse := "Hello example"
+	// tlsHandlerTLS is a HandlerFunc that can be set to expect or not TLS
+	// connection state.
+	tlsHandlerTLS := func(tlsExpected bool) func(ResponseWriter, *Msg) {
+		return func(w ResponseWriter, req *Msg) {
+			m := new(Msg)
+			m.SetReply(req)
+			tlsFound := true
+			if connState := w.(ConnectionState).ConnectionState(); connState.Version == 0 {
```

Comment:
I think `ConnectionState` has to be returned to a pointer, it doesn't seem at all idiomatic or obvious to have to compare `Version == 0`. (cc @miekg).

Reply:
I don't know why the std lib choose not to do so, but I like to mimic that as close as possible, i.e. return a non-pointer.